probe_src/performance_test.py

```diff
@@ -1,190 +1,224 @@
+    if os.path.exists(RECORD_DIR):
+        print("Removing record directory.")
+        shutil.rmtree(RECORD_DIR)
+    time.sleep(3)
+
+def get_process_stats(proc):
+    try:
+        rusage = resource.getrusage(resource.RUSAGE_CHILDREN)
+        cpu_times = (rusage.ru_utime, rusage.ru_stime)  
+        memory_usage = rusage.ru_maxrss  
+    except Exception as e:
+        print(f"Failed to get process stats: {e}")
+        cpu_times = memory_usage = None
+
+    io_counters = None
+    try:
+        process = psutil.Process(proc.pid)
+        io_counters_raw = process.io_counters()
+        io_counters = {
+            'read_count': io_counters_raw.read_count,
+            'write_count': io_counters_raw.write_count,
+            'read_bytes': io_counters_raw.read_bytes,
+            'write_bytes': io_counters_raw.write_bytes
+        }
+    except Exception as e:
+        print(f"Failed to get I/O counters: {e}")
+
+    return cpu_times, memory_usage, io_counters
+
+def benchmark_command(command: str, warmup_iterations: int, benchmark_iterations: int, flag: bool) -> list[Result]:
     results = []
 
     for _ in range(warmup_iterations):
+        print(f"Running warmup command: {command}")
+        subprocess.run("rm probe_log", shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
         subprocess.run(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
 
-    for _ in range(benchmark_iterations):
-        start_time_psutil = datetime.datetime.now()
-
-        proc = psutil.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-
-        cpu_times = memory_info = io_counters = None
-
-        if proc.is_running():
-            try:
-                cpu_times = proc.cpu_times()
-                memory_info = proc.memory_info()
-                io_counters = proc.io_counters()
-            except (psutil.NoSuchProcess, psutil.AccessDenied):
-                pass
-
-        stdout, stderr = proc.communicate()
-
-        datetime.datetime.now()
-
-        datetime.datetime.now()
-        returncode = proc.wait()
-        end_time_wait = datetime.datetime.now()
-
-        result = Result(
-            returncode=returncode,
-            cpu_times=cpu_times,
-            memory_info=memory_info,
-            io_counters=io_counters,
-            stdout=stdout.decode('utf-8'),
-            stderr=stderr.decode('utf-8'),
-            start_time=start_time_psutil,
-            end_time=end_time_wait
-        )
-        results.append(result)
-        time.sleep(4)  # Pause for 4 seconds after each command
+    if not flag:
+        for _ in range(benchmark_iterations):
+            cleanup()
+            start_time_psutil = datetime.datetime.now()
+            print(f"Starting process with command: {command}")
+            proc = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

Comment:
We should send the command as a list of strings and **not** pass `shell=True`. We would be "splitting" the command into arguments rather than asking the shell to do that.

Two reasons for this:
1. Invoking the shell adds some unrelated time and resource utilization.
2. The string is already split correctly, and asking it to be joined only to be split again could introduce errors. For example, suppose one command had a space in it: `cat hello\ world.txt`. I believe the current code would mis-interpret that as three arguments instead of two. Instead, since we already have the command split by arguments, we would say `["cat", "hello world.txt"]`, which is unambiguously two arguments.